tempesta_fw/http_match.c

```diff
@@ -201,27 +202,99 @@ match_hdr(const TfwHttpReq *req, const TfwHttpMatchRule *rule)
 	return hdr_val_eq(req, id, rule->arg.str, rule->arg.len, flags);
 }
 
+#define _MOVE_TO_COND(p, end, cond)			\
+	while((p) < (end) && !(cond)) {			\
+		(p)++;					\
+	}
+
+/* It would be hard to apply some header-specific rules here, so ignore
+ * case for all headers according to the robustness principle.
+ */
 static bool
 match_hdr_raw(const TfwHttpReq *req, const TfwHttpMatchRule *rule)
 {
-	TfwStr *hdr;
 	int i;
 	tfw_str_eq_flags_t flags = map_op_to_str_eq_flags(rule->op);
 
-	/* It would be hard to apply some header-specific rules here, so ignore
-	 * case for all headers according to the robustness principle. */
-	flags |= TFW_STR_EQ_CASEI;
-
 	for (i = 0; i < req->h_tbl->size; ++i) {
+		const TfwStr *hdr, *dup, *end, *chunk;
+		const char *c, *cend, *p, *pend;
+		char prev;
+		short cnum;
+
 		hdr = &req->h_tbl->tbl[i];
-		if (TFW_STR_EMPTY(hdr))
+		if (TFW_STR_EMPTY(hdr)) {
 			continue;
+		}
+
+		TFW_STR_FOR_EACH_DUP(dup, hdr, end) {
+			/* Initialize  the state - get the first chunk. */
+			p = rule->arg.str;
+			pend = rule->arg.str + rule->arg.len;
+			cnum = 0;
+			chunk = TFW_STR_CHUNK(dup, 0);
+			if (!chunk) {
+				return p == NULL;
+			}
+			c = chunk->ptr;
+			cend = chunk->ptr + chunk->len;
+
+#define _TRY_NEXT_CHUNK(ok_code, err_code)		\
+	if (unlikely(c == cend))	{		\
+		++cnum;					\
+		chunk = TFW_STR_CHUNK(dup, cnum); 	\
+		if (chunk) {				\
+			c = chunk->ptr;			\
+			cend = chunk->ptr + chunk->len; \
+			ok_code;			\
+		} else {				\
+			err_code;			\
+		}					\
+	}
 
-		/* TODO: handle LWS* between header and value for raw headers.
-		 * (currently "X-Hdr:foo" is not equal to "X-Hdr: foo").
-		 */
-		if (tfw_str_eq_cstr(hdr, rule->arg.str, rule->arg.len, flags))
-			return true;
+state_common:
+			while (p != pend && c != cend) {
+				if (tolower(*p) != tolower(*c)) {
+					/* If the position of the strings have a
+					 * different number of space characters,
+					 * or one of the strings have space
+					 * characters after ':', missing their
+					 */
+					if (unlikely((
+						(isspace(*p) &&
+							(prev == *p ||
+								prev == ':'))
+						||
+						(isspace(*c) &&
+							(prev == *c ||
+								prev == ':'))
+						))) {
+						goto state_sp;
+					}
+
+					return false;
+				}
+
+				p++;
+				c++;
+				prev = *(p - 1);
+			}
+			_TRY_NEXT_CHUNK(goto state_common, );
+
+			if (c == cend && p == pend) {
+				return true;
+			}
+
+			if (p == pend && flags & TFW_STR_EQ_PREFIX) {
+				return true;
+			}
+
+state_sp:
+			_MOVE_TO_COND(p, pend, !isspace(*p));
+			_MOVE_TO_COND(c, cend, !isspace(*c));
+			_TRY_NEXT_CHUNK(goto state_sp, return p == pend);
+			goto state_common;
```

Comment:
Case 1: we're from while() by goto statement. Since while() check passed, we know that neither c nor p reach end of their strings, so why do we try the next chunk? At least one unnecessary condition, but also unclear code.

Case 2: we're here due to end of pattern - what's the sense to try the next chunk?

Code between lines 282-295 do 3 checks for ```c == end``` - this has no sense.

I'd say you're trying to generalize different cases and pay by unnecessary checks and conditional jumps (which are expensive) for the generalization.